test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go

```diff
@@ -0,0 +1,459 @@
+	)
+}
+
+// DockerClusterToDockerMachines is a handler.ToRequestsFunc to be used to enqueue
+// requests for reconciliation of DockerMachines.
+func (r *DockerMachinePoolReconciler) DockerClusterToDockerMachinePools(o handler.MapObject) []ctrl.Request {
+	result := []ctrl.Request{}
+	c, ok := o.Object.(*infrav1.DockerCluster)
+	if !ok {
+		r.Log.Error(errors.Errorf("expected a DockerCluster but got a %T", o.Object), "failed to get DockerMachinePool for DockerCluster")
+		return nil
+	}
+	log := r.Log.WithValues("DockerCluster", c.Name, "Namespace", c.Namespace)
+
+	cluster, err := util.GetOwnerCluster(context.TODO(), r.Client, c.ObjectMeta)
+	switch {
+	case apierrors.IsNotFound(err) || cluster == nil:
+		return result
+	case err != nil:
+		log.Error(err, "failed to get owning cluster")
+		return result
+	}
+
+	labels := map[string]string{clusterv1.ClusterLabelName: cluster.Name}
+	machinePoolList := &clusterv1exp.MachinePoolList{}
+	if err := r.Client.List(context.TODO(), machinePoolList, client.InNamespace(c.Namespace), client.MatchingLabels(labels)); err != nil {
+		log.Error(err, "failed to list DockerMachinePools")
+		return nil
+	}
+
+	for _, m := range machinePoolList.Items {
+		if m.Spec.Template.Spec.InfrastructureRef.Name == "" {
+			continue
+		}
+		name := client.ObjectKey{Namespace: m.Namespace, Name: m.Name}
+		result = append(result, ctrl.Request{NamespacedName: name})
+	}
+
+	return result
+}
+
+func (r *DockerMachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, machinePool *clusterv1exp.MachinePool, dockerMachinePool *infrav1exp.DockerMachinePool, log logr.Logger) (ctrl.Result, error) {
+	// check if there are more machines than the spec requires
+	realMachineCount := int32(len(dockerMachinePool.Status.Machines))
+	if realMachineCount > 0 && realMachineCount > *machinePool.Spec.Replicas {
+		for i := *machinePool.Spec.Replicas; i < realMachineCount; i++ {
+			if err := r.reconcileDockerWorkerMachine(ctx, cluster.Name, i, machinePool, dockerMachinePool, log); err != nil {
+				return ctrl.Result{}, errors.Wrapf(err, "failed to delete over-provisioned docker worker machine")
+			}
+		}
+	}
+
+	controllerutil.RemoveFinalizer(dockerMachinePool, infrav1exp.MachinePoolFinalizer)
+	return ctrl.Result{}, nil
+}
+
+func (r *DockerMachinePoolReconciler) reconcileNormal(ctx context.Context, cluster *clusterv1.Cluster, machinePool *clusterv1exp.MachinePool, dockerMachinePool *infrav1exp.DockerMachinePool, log logr.Logger) (ctrl.Result, error) {
+	// Make sure bootstrap data is available and populated.
+	if machinePool.Spec.Template.Spec.Bootstrap.DataSecretName == nil {
+		log.Info("Waiting for the Bootstrap provider controller to set bootstrap data")
+		return ctrl.Result{}, nil
+	}
+
+	if machinePool.Spec.Replicas == nil {
+		machinePool.Spec.Replicas = pointer.Int32Ptr(1)
+	}
+
+	needToRequeue := false
+	for i := int32(0); i < *machinePool.Spec.Replicas; i++ {
```

Comment:
I think that we should always read the real state from docker, and then decide if to create or to delete machine, e.g.
- list the current real machines (that are the containers controlled by the MachinePool, detected by naming convention or using labels)
- reconcile machines that should exist
- delete machines that should not exist

Reply:
Agreed, and updated to reflect this feedback.